certstore_linux.go

```diff
@@ -1,14 +1,318 @@
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// CertificateChain implements the Identity interface.
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	var (
+		cert  = C.CERTCertificate(*i)
+		certs = C.CERT_GetCertChainFromCert(&cert, C.PR_Now(), C.certUsageAnyCA)
+	)
+	if certs == nil {
+		return nil, fmt.Errorf("error building certificate chain: %s", C.GoString(C.GetErrorString()))
+	}
+	defer C.CERT_DestroyCertList(certs)
+
+	var (
+		certificates = make([]*x509.Certificate, 0)
+		node         *C.CERTCertListNode
+	)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		var (
+			identity  = nssIdentity(*C.CERT_DupCertificate(node.cert))
+			cert, err = identity.Certificate()
+		)
+		if err != nil {
+			return nil, errors.New("unable to parse certificate when building chain")
+		}
+		certificates = append(certificates, cert)
+	}
+	return certificates, nil
+}
+
+// Delete implements the Identity interface.
+func (i *nssIdentity) Delete() error {
+	var (
+		cert      = C.CERTCertificate(*i)
+		secstatus = C.PK11_DeleteTokenCertAndKey(&cert, nil)
+	)
+	if secstatus != C.SECSuccess {
+		return errors.New("failed to delete certificate & key pair")
+	}
+	return nil
+}
+
+// Close implements the Store interface.
+func (i *nssIdentity) Close() {
+	cert := C.CERTCertificate(*i)
+	C.CERT_DestroyCertificate(&cert)
+}
+
+// Public implements the crypto.Signer interface.
+func (i *nssIdentity) Public() crypto.PublicKey {
+	var cert, _ = i.Certificate()
+	if cert == nil {
+		return nil
+	}
+	return cert.PublicKey
 }
 
-// Implement this function, just to silence other compiler errors.
+// Sign implements the crypto.Signer interface.
+func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash := opts.HashFunc()
+	if len(digest) != hash.Size() {
+		return nil, errors.New("bad digest for hash")
+	}
+	var (
+		cert = C.CERTCertificate(*i)
+		key  = C.PK11_FindKeyByAnyCert(&cert, nil)
+	)
+	if key == nil {
+		return nil, errors.New("cannot find private key")
+	}
+
+	digestC := C.SECITEM_AllocItem(nil, nil, C.uint(len(digest)))
+	if digestC == nil {
+		return nil, errors.New("failure to allocate memory for digest")
+	}
+	defer C.SECITEM_FreeItem(digestC, 1)
+
+	resultC := C.SECITEM_AllocItem(nil, nil, C.uint(C.PK11_SignatureLen(key)))
+	if resultC == nil {
+		return nil, errors.New("failure to allocate memory for signature")
+	}
+	defer C.SECITEM_FreeItem(resultC, 1)
+
+	digestC.len = C.uint(len(digest))
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+
+	if C.SGN_Digest(key, goToNssAlg[hash], resultC, digestC) != C.SECSuccess {
+		if C.PORT_GetError() == C.SEC_ERROR_INVALID_ALGORITHM {
+			return nil, ErrUnsupportedHash
+		}
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
+	}
+
+	if key.keyType == C.ecKey {
```

Comment:
We have to do this same conversion for Windows [here](https://github.com/mastahyeti/certstore/blob/bad86412360c8210ee584cdeb70e4d929ff48f40/certstore_windows.go#L419-L438). It might be nicer to extract that logic into a shared helper method and use it here as well to save another cgo call.